File: cpp/src/gridmap.cpp
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <queue>
#include "gridmap.h"

#include <cairo.h>
#include <boost/random/mersenne_twister.hpp>
#include <boost/random/uniform_int_distribution.hpp>

using std::cout;
using std::endl;

GridMap::GridMap() {}
GridMap::GridMap(const int xmax, const int ymax) 
{
	GridCell blank_cell;
	mMap.resize(xmax);
	for (int i = 0; i < xmax; i++) {
		mMap[i].resize(ymax);
		for (int j = 0; j < ymax; j++) {
			mMap[i][j] = blank_cell;
		}
	}
}

void GridMap::setCell(const int x, const int y, const GridCell cell)
{
	mMap[x][y] = cell;
}

GridCell* GridMap::getCell(const int x, const int y)
{
	return &(mMap[x][y]);
}

int GridMap::getSizeX() {return static_cast<int>(mMap.size());}
int GridMap::getSizeY()
{
	if (this->getSizeX() > 0) {
		return static_cast<int>(mMap[0].size());
	} else {
		return 0;
	}
}

void GridMap::placeRoomDimensions(const int orig_x,const int orig_y,const int wall_h,const int wall_v)
{
	if (orig_x < 0 || orig_y < 0) {
		std::cerr << "ERROR: room origin coordinates are outside of the map." << std::endl;
		std::cerr << "\t Coordinates: ("<< orig_x << ", " << orig_y << ")." << std::endl;
		return;
	}

	int max_x = this->getSizeX();
	int max_y = this->getSizeY();

	if (	orig_x+wall_h > max_x ||
		orig_x+wall_h < 0 ||
		orig_y+wall_v > max_y ||
		orig_y+wall_v < 0 )
	{
		std::cerr << "ERROR: room boundaries are outside of the map." << std::endl;
		return;
	}

	int lower_x, upper_x;
	int lower_y, upper_y;

	if (wall_h > 0) {
		lower_x = orig_x;
		upper_x = orig_x+wall_h;
	} else {
		lower_x = orig_x+wall_h;
		upper_x = orig_x;
	}

	if (wall_v > 0) {
		lower_y = orig_y;
		upper_y = orig_y+wall_v;
	} else {
		lower_y = orig_y+wall_v;
		upper_y = orig_y;
	}

	for (int i = lower_x; i < upper_x; i++) {
		for (int j = lower_y; j < upper_y; j++) {
			if (this->mMap[i][j].isEntrance() == false) {
				this->mMap[i][j] = GridCell(Area_Type::Room);
			}
		}
	}
}

void GridMap::placeRoom(const int x0, const int y0, const int x1, const int y1)
{
	int lower_x, lower_y, upper_x, upper_y;

	if (x1 > x0) {
		lower_x = x0;
		upper_x = x1;
	} else {
		lower_x = x1;
		upper_x = x0;
	}

	if (y1 > y0) {
		lower_y = y0;
		upper_y = y1;
	} else {
		lower_y = y1;
		upper_y = y0;
	}

	if (lower_x < 0) {
		lower_x = 0;
	}

	if (lower_y < 0) {
		lower_y = 0;
	}

	if (upper_x >= this->getSizeX()) {
		upper_x = this->getSizeX() - 1;
	}

	if (upper_y >= this->getSizeY()) {
		upper_y = this->getSizeY() - 1;
	}

	for (int i = lower_x; i <= upper_x; i++) {
		for (int j = lower_y; j <= upper_y; j++) {
			this->mMap[i][j] = GridCell(Area_Type::Room);
		}
	}
}

void GridMap::placeHallway(const int orig_x, const int orig_y, const int dest_x, const int dest_y, const int width, const Route_t route) 
{
	static int salt = 0;
	Route_t route_type = route;

	if (route == Route_t::Manhattan) {
		salt += 1;
		boost::random::mt19937 gen(std::time(0)+salt);
		boost::random::uniform_int_distribution<> dist(0,1);

		int num = dist(gen);

		cout << num << endl;

		if (num == 0) {
			route_type = Route_t::Horizontal_First;
		} else {
			route_type = Route_t::Vertical_First;
		}
	} else if (route != Route_t::Horizontal_First && route != Route_t::Vertical_First) {
		cout << "WARNING: Only Manhattan routing has been implemented." << endl;
	}

	if (route_type == Route_t::Horizontal_First) {
		this->placeRoom(orig_x,orig_y,dest_x,orig_y);
		this->placeRoom(dest_x,orig_y,dest_x,dest_y);
	} else if (route_type == Route_t::Vertical_First) {
		this->placeRoom(orig_x,orig_y,orig_x,dest_y);
		this->placeRoom(orig_x,dest_y,dest_x,dest_y);
	} else {
		cout << "WARNING: Hallway not placed due to lack of implementation..." << endl;
	}
}

void GridMap::placeRandomRoom(const int scale, bool connected)
{
	// Salt the RNG so that multiple calls generate different numbers.
	static int salt = 0;
	salt += 1;

	// Set up random number generator
	boost::random::mt19937 gen(std::time(0)+salt);
	boost::random::uniform_int_distribution<> dist(2,scale);

	// Generate width and height of the room.
	int width = dist(gen);
	int height = dist(gen);

	// Generate location of room.
	boost::random::uniform_int_distribution<> orig_x(1,this->getSizeX());
	boost::random::uniform_int_distribution<> orig_y(1,this->getSizeY());
	int x0 = orig_x(gen);
	int y0 = orig_y(gen);

	if (connected) {
		int xorig = x0+width/2;
		int yorig = y0+height/2;
		if (xorig > this->getSizeX() || yorig > this->getSizeY()) {
			return;
		}

		int x, y;
		this->findNearestConnected(xorig,yorig,x,y);
		this->placeHallway(xorig,yorig,x,y);
		cout << "Room coords (" << xorig << ", "<<yorig << ")"<< endl;
		cout << "Connect     (" << x << ", " << y << ")" << endl;
	}

	// Place room on the map.
	this->placeRoom(x0-width/2,y0-height/2,x0+width/2,y0+height/2);
}

void GridMap::findNearestConnected(const int x0, const int y0, int& x, int& y)
{
	std::queue< std::vector<int> > rooms;
	int radius = 0;
	bool found = false;

	while (!found) {
		if (found) {
			cout << "How the fuck did I get here..." << endl;
		}

		radius += 1;

		int xmin = x0-radius;
		int xmax = x0+radius;
		int ymin = y0-radius;
		int ymax = y0+radius;

		if (xmin < 0) {
			xmin = 0;
		}
		
		if (ymin < 0) {
			ymin = 0;
		}

		if (xmax >= this->getSizeX()) {
			xmax = this->getSizeX()-1;
		}

		if (ymax >= this->getSizeY()) {
			ymax = this->getSizeY()-1;
		}

		if (	xmin == 0 
			&& ymin == 0 
			&& xmax == (this->getSizeX()-1) 
			&& ymax == (this->getSizeY()-1))
		{
			//TODO Throw exception instead.
			found = true;
			break;
		}

		// Scan bottom X
		for (int i = xmin; i <= xmax; i++) {
			if ((this->getCell(i,ymin))->isRoom()) {
				found = true;
				std::vector<int> coord(2);
				coord[0] = i;
				coord[1] = ymin;
				rooms.push(coord);
				// NOTE: Do not stop, just keep going and add to rooms queue.
			}
		}

		// Scan for top X
		for (int i = xmin; i <= xmax; i++) {
			if ((this->getCell(i,ymax))->isRoom()) {
				found = true;
				std::vector<int> coord(2);
				coord[0] = i;
				coord[1] = ymax;
				rooms.push(coord);
				// NOTE: Do not stop, just keep going and add to rooms queue.
			}
		}

		// Scan for left Y
		for (int i = ymin; i <= ymax; i++) {
			if ((this->getCell(xmin,i))->isRoom()) {
				found = true;
				std::vector<int> coord(2);
				coord[0] = xmin;
				coord[1] = i;
				rooms.push(coord);
				// NOTE: Do not stop, just keep going and add to rooms queue.
			}
		}

		// Scan right left Y
		for (int i = ymin; i <= ymax; i++) {
			if ((this->getCell(xmax,i))->isRoom()) {
				found = true;
				std::vector<int> coord(2);
				coord[0] = xmax;
				coord[1] = i;
				rooms.push(coord);
				// NOTE: Do not stop, just keep going and add to rooms queue.
			}
		}

	}

	if (rooms.size() > 0) {
		static int salt = 0;
		salt += 1;
	
		boost::random::mt19937 gen(std::time(0)+salt);
		boost::random::uniform_int_distribution<> dist(0,rooms.size()-1);
	
		int choice = dist(gen);
	
		for (int i = 0; i < choice; i++) {
			rooms.pop();
		}
	
		// These are the actual coordinates that are returned.
		x = (rooms.front())[0];
		y = (rooms.front())[1];
	} else {
		x = 0;
		y = 0;
	}
}


void GridMap::placeEntrance(const int x, const int y) {
	this->mMap[x][y].setArea(Area_Type::Entrance);
}

void GridMap::generate_random_cells(int limit)
{
	GridCell room_cell(Area_Type::Room);
	// Create the random-number generator and seed with the current time.
	boost::random::mt19937 gen(std::time(0));
	boost::random::uniform_int_distribution<> dist(1,100);
	for (int i = 0; i < this->getSizeX(); i++) {
		for (int j = 0; j < this->getSizeY(); j++) {
			int val = dist(gen);
			if (val >= limit) this->setCell(i,j,room_cell);
		}
	}
}


void GridMap::generate_annealed_random_cells()
{
	GridCell empty_cell(Area_Type::Nothing);
	// Start by generating a random grid.
	this->generate_random_cells(80);
	
	// Anneal by removing stragglers.
	for (int i = 1; i < this->getSizeX()-1; i++) {
		for (int j = 1; j < this->getSizeY()-1; j++) {
			bool alone = mMap[i-1][j].isEmpty() 
				&& mMap[i][j-1].isEmpty() 
				&& mMap[i+1][j].isEmpty() 
				&& mMap[i][j+1].isEmpty();
			if (alone) this->setCell(i,j,empty_cell);
		}
	}
}

void GridMap::draw(cairo_t* ctx) {
	// Use a default of 10px for the map scale.
	this->draw(ctx,10);
}

void GridMap::draw(cairo_t* ctx, const int room_scale)
{
	int max_x = this->getSizeX();
	int max_y = this->getSizeY();

	// TEST
	// Load cobblestone floor
	cairo_surface_t* floor = cairo_image_surface_create_from_png("floor_light.png");

	for (int i = 0; i < max_x; i++) {
		for (int j = 0; j < max_y; j++) {
			GridCell& cell = *(this->getCell(i,j));
			if (cell.isRoom()) {
				// Calculate rectangle coordinates
				int x1 = i*room_scale;
				int y1 = j*room_scale;

				// Draw cell
				cairo_set_source_rgb(ctx,0.8,0.8,0.8);
				cairo_rectangle(ctx,x1,y1,room_scale,room_scale);
				cairo_fill(ctx);
				//
				//TEST cobblestone floor.
			//	cairo_set_source_surface(ctx,floor,x1,y1);
			//	cairo_paint(ctx);
			} else if (cell.isEntrance()) {
				int x1 = i*room_scale;
				int y1 = j*room_scale;

				// Set to bright pink
				cairo_set_source_rgb(ctx,1.0,0.078,0.5764);
				cairo_rectangle(ctx,x1,y1,room_scale,room_scale);
				cairo_fill(ctx);
			} else if (cell.getArea() == Area_Type::Tested) {
				int x1 = i*room_scale;
				int y1 = j*room_scale;

				// Set to bright pink
				cairo_set_source_rgb(ctx,0,0,1.0);
				cairo_rectangle(ctx,x1,y1,room_scale,room_scale);
				cairo_fill(ctx);
			}

			if (cell.getVertWall() == Wall_Type::Wall) {
				int x1 = i*room_scale;
				int y1 = j*room_scale;

				// Draw wall
				//TODO
			}

			if (cell.getHorizWall() == Wall_Type::Wall) {
				int x1 = i*room_scale;
				int y1 = j*room_scale;

				//Draw wall
				//TODO
			}
		}
	}
}

bool GridMap::cave_anneal_cell(const int i, const int j)
{
	int neighbours = 0;

	bool onEdge = this->onEdge(i,j);

	if (onEdge == false) {
		neighbours += (this->getCell(i-1,j-1))->isRoom() ? 1 : 0;
		neighbours += (this->getCell(i  ,j-1))->isRoom() ? 1 : 0;
		neighbours += (this->getCell(i+1,j-1))->isRoom() ? 1 : 0;
		neighbours += (this->getCell(i-1,j  ))->isRoom() ? 1 : 0;
		neighbours += (this->getCell(i+1,j  ))->isRoom() ? 1 : 0;
		neighbours += (this->getCell(i-1,j+1))->isRoom() ? 1 : 0;
		neighbours += (this->getCell(i  ,j+1))->isRoom() ? 1 : 0;
		neighbours += (this->getCell(i+1,j+1))->isRoom() ? 1 : 0;

		// Not technically a neighbour, but simplifies the conditional.
		neighbours += (this->getCell(i,j))->isRoom() ? 1 : 0;

		if (neighbours >= 5) {
			return true;
		} else {
			return false;
		}
	} else {
		return false;
	}
}

void GridMap::generate_cave_iteration()
{
	GridMap tmp_map(this->getSizeX(),this->getSizeY());
	GridCell cave_cell(Area_Type::Room);
	GridCell rock_cell(Area_Type::Nothing);

	for (int i = 0; i < tmp_map.getSizeX(); i++) {
		for (int j = 0; j < tmp_map.getSizeY(); j++) {
			if (cave_anneal_cell(i,j) == true) {
				tmp_map.setCell(i,j,cave_cell);
			} else {
				tmp_map.setCell(i,j,rock_cell);
			}
		}
	}

	(*this) = tmp_map;
}

void GridMap::generate_cave(int num_iterations,int seed_limit)
{
	this->generate_random_cells(seed_limit);
	for (int i = 0; i < num_iterations; i++) {
		this->generate_cave_iteration();
	}
}

bool GridMap::onEdgeX(const int x) {
	if ( x == 0 || x == this->getSizeX()-1) {
		return true;
	} else {
		return false;
	}
}

bool GridMap::onEdgeY(const int y) {
	if ( y == 0 || y == this->getSizeY()-1) {
		return true;
	} else {
		return false;
	}
}

bool GridMap::onEdge(const int x, const int y) {
	return onEdgeX(x) || onEdgeY(y);
}

void GridMap::create_test_map()
{
	// 1x1 room at 1.
	this->placeRoom(1,1,1,1);

	// 4x4 room at 3,1
	this->placeRoom(3,1,4,4);

	// Create a set of hallways:
	this->placeHallway(8,1,11,4,1,Route_t::Horizontal_First);
	this->placeHallway(13,1,16,4,1,Route_t::Vertical_First);
	this->placeHallway(18,4,21,1,1,Route_t::Horizontal_First);
	this->placeHallway(23,4,26,1,1,Route_t::Vertical_First);
	this->placeHallway(31,1,28,4,1,Route_t::Horizontal_First);
	this->placeHallway(36,1,33,4,1,Route_t::Vertical_First);
}

File: python/rpgmap/gridmap.py
import collections
import copy
import random

from .route import Route_t


class Area_Type:
    Nothing = 0
    Entrance = 1
    Room = 2
    Stairs = 3
    Tested = 4

class Wall_Type:
    Nothing = 0
    Wall = 1
    Door = 2
    Secret_Door = 3

class Point_Type:
    Nothing = 0
    Pillar = 1

class Accessable_t:
    Yes = True
    No = False
    Unknown = None


class GridCell:
    def __init__(self, area=Area_Type.Nothing, horizontal_wall=Wall_Type.Nothing, 
                       vertical_wall=Wall_Type.Nothing, point=Point_Type.Nothing, 
                       access=None):
        self.accessable = access
        self.areaType = area
        self.verticalWallType = vertical_wall
        self.horizontalWallType = horizontal_wall
        self.pointType = point

    @property
    def isAccessable(self):
        return bool(self.accessable)

    @property
    def isEmpty(self):
        return (self.areaType == Area_Type.Nothing and
                self.verticalWallType == Wall_Type.Nothing and
                self.horizontalWallTYpe == Wall_Type.Nothing and
                self.pointType == Point_Type.Nothing)

    @property
    def isEntrance(self):
        return self.areaType == Area_Type.Entrance

    @property
    def isRoom(self):
        return self.areaType == Area_Type.Room

    def setArea(self, areaType):
        self.areaType = areaType





class GridMap:
    def __init__(self, xmax=None, ymax=None):
        self._xmax = xmax
        self._ymax = ymax

        if self._xmax and self._ymax:
            self._cells = [[GridCell() for y in range(self._ymax)] for x in range(self._xmax)]


    def setCell(self, x, y, cell):
        self._cells[x][y] = cell

    def getCell(self, x, y):
        return self._cells[x][y]

    @property
    def sizeX(self):
        return self._xmax

    @property
    def sizeY(self):
        return self._ymax

    def placeRoom(self, x0, y0, x1, y1):
        # Ensure that the order is well-formed 
        lower_x, upper_x = sorted((x0, x1))
        lower_y, upper_y = sorted((y0, y1))

        # Clamp the lower bounds to 0
        lower_x = max(0, lower_x)
        lower_y = max(0, lower_y)
            
        # Clamp the upper bounds to max_x, max_y
        upper_x = min(self.sizeX, upper_x)
        upper_y = min(self.sizeY, upper_y)

        for i in range(lower_x, upper_x):
            for j in range(lower_y, upper_y):
                self._map[i][j].areaType = Area_Type.Room

    def placeRoomDimensions(self, orig_x, orig_y, wall_h, wall_v):
        if orig_x < 0 or orig_y < 0:
            print("ERROR: room origin coordinates are outside the map\n"
                  f"\tCoordinates: ({orig_x}, {orig_y}).", file=sys.stderr)
            return
        
        max_x = self.sizeX
        max_y = self.sizeY

        if not (0 < orig_x+wall_h < max_x) or not (0 < orig_y+wall_v < max_y):
            print("ERROR: room boundaries are outside the map", file=sys.stderr)
            return

        if wall_h > 0:
            lower_x = orig_x
            upper_x = orig_x+wall_h
        else:
            lower_x = orig_x+wall_h
            upper_x = orig_x

        if wall_v > 0:
            lower_y = orig_y
            upper_y = orig_y+wall_v
        else:
            lower_y = orig_y+wall_v
            upper_y = orig_y

        for i in range(lower_x, upper_x):
            for j in range(lower_y, upper_y):
                if not self._map[i][j].isEntrace:
                    self._map[i][j] = GridCell(Area_Type.room)

    def placeHallway(self, orig_x, orig_y, dest_x, dest_y, width=1, route=Route_t.Manhattan):
        route_type = route

        if route == Route_t.Manhattan:
            num = random.randint(0, 1)
            print(num)
            if num == 0:
                route_type = Route_t.Horizontal
            else:
                route_type = Route_t.Vertical
        elif route not in {Route_t.Horizontal, Route_t.Vertical}:
            print("WARNING: Only Manhattan routing has been implemented.", file=sys.stderr)

        if route_type == Route_t.Horizontal:
            self.placeRoom(orig_x, orig_y, dest_x, orig_y)
            self.placeRoom(dest_x, orig_y, dest_x, dest_y)
        else:
            self.placeRoom(orig_x, orig_y, orig_x, dest_y)
            self.placeRoom(orig_x, dest_y, dest_x, dest_y)


    def placeRandomRoom(self, scale, connected=True):
        width = random.randint(2, scale)
        height = random.randint(2, scale)

        x0 = random.randint(1, self.sizeX)
        x1 = random.randint(1, self.sizeY)

        if connected:
            xorig = x0+width//2
            yorig = y0+height//2
            if xorig > self.sizeX or yorig > self.sizeY:
                return
            (x, y) = self.findNearestConnected(xorig, yorig)
            self.placeHallway(xorig, yorig, x, y)
            print(f"Room coords ({xorig}, {yorig})\n"
                  f"Connect     ({x}, {y})")

            self.placeRoom(x0-width//2, y0-height//2, x0+width//2, y0+height//2)


    def placeEntrance(self, x, y):
        # Do in a for loop so that we can keep looking if the first one isn't a room
        for xShift in range(x, self.sizeX):
            if self._cells[xShift][y].isRoom:
                self._cells[xShift][y].setArea(Area_Type.Entrance)
                break

    def findNearestConnected(self, x0, y0):
        rooms = collections.deque()
        radius = 0
        found = False

        while not found:
            radius += 1
            xmin = x0-radius
            xmax = x0+radius
            ymin = y0-radius
            ymax = y0+radius

            if xmin < 0:
                xmin = 0
            if ymin < 0:
                ymin = 0

            if xmax >= self.sizeX:
                xmax = self.sizeX - 1
            if ymax >= self.sizeY:
                ymax = self.sizeY - 1

            if xmin == 0 and ymin == 0 and xmax == self.sizeX-1 and ymax == self.sizeY-1:
                # Todo: exception or other?
                found = True

            # Scan bottom x
            for i in range(xmin, xmax+1):
                if self.getCell(i, ymax).isRoom:
                    found = True
                    coord = (i, ymax)
                    rooms.append(coord)
                    # NOTE: Do not stop, just keep going and add to rooms queue

            # Scan top x
            for i in range(xmin, xmax+1):
                if self.getCell(i, ymin).isRoom:
                    found = True
                    coord = (i, ymin)
                    rooms.append(coord)
                    # NOTE: Do not stop, just keep going and add to rooms queue

            # Scan for left Y
            for i in range(ymin, ymax+1):
                if self.getCell(xmin, i).isRoom:
                    found = True
                    coord = (xmin, i)
                    rooms.append(coord)
                    # NOTE: Do not stop, just keep going and add to rooms queue

            # Scan for right Y
            for i in range(ymin, ymax+1):
                if self.getCell(xmax, i).isRoom:
                    found = True
                    coord = (xmax, i)
                    rooms.append(coord)
                    # NOTE: Do not stop, just keep going and add to rooms queue

        if rooms:
            choice = random.randint(0, len(rooms)-1)
            return rooms[choice]
        return (0, 0)


    # Limit is a value between 1 and 100. This limit sets the change that the cells are 
    # a room. (Higher means fewer rooms).
    def generate_random_cells(self, limit):
        for i in range(self.sizeX):
            for j in range(self.sizeY):
                val = random.randint(1, 100)
                if val > limit:
                    self.setCell(i, j, GridCell(Area_Type.Room))

    def generate_annealed_random_cells(self):
        # Start by generating a random grid
        self.generate_random_cells(80)

        # Anneal by removing stragglers
        for i in range(1, self.sizeX-1):
            for j in range(1, self.sizeY-1):
                alone = self._map[i-1][j].isEmpty and self._map[i][j-1].isEmpty and self._map[i+1][j].isEmpty and self._map[i][j+1].isEmpty
                if alone:
                    self.setCell(i, j, empty_cell)


    def generate_cave(self, num_iterations, seed_limit=55):
        self.generate_random_cells(seed_limit)
        for i in range(num_iterations):
            self.generate_cave_iteration()
        self.removeOrphans()


    def cave_anneal_cell(self, i, j):
        if not self.onEdge(i, j):
            # Count the number of neighbours
            neighbours = 0
            neighbours += 1 if self.getCell(i-1, j-1).isRoom else 0
            neighbours += 1 if self.getCell(i  , j-1).isRoom else 0
            neighbours += 1 if self.getCell(i+1, j-1).isRoom else 0
            neighbours += 1 if self.getCell(i-1, j  ).isRoom else 0
            neighbours += 1 if self.getCell(i+1, j  ).isRoom else 0
            neighbours += 1 if self.getCell(i-1, j+1).isRoom else 0
            neighbours += 1 if self.getCell(i  , j+1).isRoom else 0
            neighbours += 1 if self.getCell(i+1, j+1).isRoom else 0

            # Technically not a neighbours, but simplifies the conditional.
            neighbours += 1 if self.getCell(i,j).isRoom else 0
            return neighbours >= 5
        return False


    def generate_cave_iteration(self):
        tmp_map = copy.deepcopy(self._cells)

        for i in range(self.sizeX):
            for j in range(self.sizeY):
                if self.cave_anneal_cell(i, j):
                    tmp_map[i][j].areaType = Area_Type.Room
                else:
                    tmp_map[i][j].areaType = Area_Type.Nothing

        self._cells = tmp_map
        return self._cells

    def removeOrphans(self):
        """Remove orphaned rooms from the map"""
        for x in range(self.sizeX):
            for y in range(self.sizeY):
                size = self.getRoomSize(x, y)
                if 0 < size < 15:
                    self.clearRoom(x, y)

    def getRoomSize(self, x, y):
        """Get the size of the room (number of cells) connected to (x,y)"""

        # Output value
        size = 0

        # Process all of the points in a list. Doing this in an iterative fashion because of
        # max recursion limits
        processingQueue = collections.deque([(x,y)])

        # Mask to make sure that we don't revisit rooms
        visited = [[False for y in range(self.sizeY)] for x in range(self.sizeX)]

        # Keep going while we have rooms to process
        while processingQueue:
            # Remove a room from the queue, unpack the values
            currentCell = processingQueue.pop()
            xcur, ycur = currentCell

            # If we have visited this cell before or it is not a "Room" then we should
            # just stop processing and move on to the next.
            if visited[xcur][ycur] or self._cells[xcur][ycur].areaType != Area_Type.Room:
                continue

            # We got here so it's a room that we haven't visited. Mark it as visited now
            visited[xcur][ycur] = True

            # This is a room and we haven't visited here before so add one to size
            size += 1

            # Add all neighbours to the queue
            processingQueue.extend([(xcur-1, ycur),
                                    (xcur+1, ycur),
                                    (xcur, ycur-1),
                                    (xcur, ycur+1)])
        return size
        

    def clearRoom(self, x, y):
        if self._cells[x][y].areaType == Area_Type.Nothing:
            return

        processingQueue = collections.deque([(x, y)])

        while processingQueue:
            xcur, ycur = processingQueue.pop()
            if self._cells[xcur][ycur].areaType != Area_Type.Nothing:
                self._cells[xcur][ycur].areaType = Area_Type.Nothing
                processingQueue.extend([(xcur+1, ycur),
                                        (xcur-1, ycur),
                                        (xcur, ycur+1),
                                        (xcur, ycur-1)])
        return




    def draw(self, ctx, scale=10):
        max_x = self.sizeX
        max_y = self.sizeY

        for i in range(max_x):
            for j in range(max_y):
                cell = self.getCell(i, j)

                if cell.isRoom:
                    x1 = i*scale
                    y1 = j*scale

                    ctx.set_source_rgb(0.8, 0.8, 0.8)
                    ctx.rectangle(x1, y1, scale, scale)
                    ctx.fill()
                elif cell.isEntrance:
                    x1 = i*scale
                    y1 = j*scale

                    ctx.set_source_rgb(1.0, 0.078, 0.5764)
                    ctx.rectangle(x1, y1, scale, scale)
                    ctx.fill()
                elif cell.areaType == Area_Type.Tested:
                    x1 = i*scale
                    y1 = i*scale

                    ctx.set_source_rgb(0, 0, 1.0)
                    ctx.rectangle(x1, y1, scale, scale)
                    ctx.fill()

    def onEdge(self, x, y):
        return x == 0 or x == self.sizeX-1 or y == 0 or y == self.sizeY-1
